File: 410-split-array-largest-sum/410-split-array-largest-sum.cpp
class Solution {
public:
    vector<int> ps;
    map<pair<int,int>,int> dp;
    int n;
    int solve(int idx, int m){
        if(dp.count({idx,m})) return dp[{idx,m}];
        if(m==1){
            return ps[n]-ps[idx];
        } 
        int tempans = ps[n];
        for(int endAt=idx;endAt< n-m+1;endAt++){
            auto temp = ps[endAt+1]-ps[idx];
            tempans = min(tempans,max(
                solve(endAt+1,m-1),
                temp
            ));
            if (temp>tempans){
                break;
            }
        }
        
        return dp[{idx,m}]=tempans;
    }
    
    int splitArray(vector<int>& nums, int m) {
        ps = {0};
        dp = map<pair<int,int>,int>();
        int s = 0;
        n = nums.size();
        for(int i:nums){
            s+=i;
            ps.push_back(s);
        }
        
        return solve(0,m);
    }
};

File: 410-split-array-largest-sum/410-split-array-largest-sum.py
class Solution:
    def splitArray(self, nums: List[int], m: int) -> int:
        n=len(nums)
        prefixSum = [0]+list(itertools.accumulate(nums))
        # print(prefixSum)
        
        @cache
        def solve(idx,subA):
            if subA == 1:
                return prefixSum[n]-prefixSum[idx]
            tempans = prefixSum[n]
            
            for endCurAt in range(idx,n-(subA-1)):
                firstSum = prefixSum[endCurAt+1]-prefixSum[idx]
                tempans = min(
                    tempans,
                      max(
                        firstSum,
                        solve(endCurAt+1,subA-1)
                      )
                )
                if firstSum>tempans:
                    break
            return tempans
        
        return solve(0,m)